scrapy_poet/registry.py

```diff
@@ -0,0 +1,150 @@
+
+        # The consume_modules() must be called first if you need to properly
+        # import rules from other packages. Otherwise, it can be omitted.
+        # More info about this caveat on web-poet docs.
+        consume_modules("external_package_A.po", "another_ext_package.lib")
+        SCRAPY_POET_RULES = default_registry.get_rules()
+
+    Make sure to call :py:func:`web_poet.rules.consume_modules` beforehand.
+    More info on this at web-poet_.
+    """
+
+    @classmethod
+    def from_crawler(cls, crawler: Crawler) -> Crawler:
+        if "SCRAPY_POET_OVERRIDES" in crawler.settings:
+            msg = (
+                "The SCRAPY_POET_OVERRIDES setting is deprecated. "
+                "Use SCRAPY_POET_RULES instead."
+            )
+            warn(msg, DeprecationWarning, stacklevel=2)
+            rules = crawler.settings.getlist("SCRAPY_POET_OVERRIDES", [])
+        else:
+            rules = crawler.settings.getlist("SCRAPY_POET_RULES", [])
+        return cls(rules=rules)
+
+    def __init__(self, rules: Optional[Iterable[ApplyRule]] = None) -> None:
+        super().__init__(rules=rules)
+        self.overrides_matcher: Dict[Type[ItemPage], URLMatcher] = defaultdict(
+            URLMatcher
+        )
+        self.item_matcher: Dict[Optional[Type[Any]], URLMatcher] = defaultdict(
+            URLMatcher
+        )
+        for rule_id, rule in enumerate(self._rules):
+            self.add_rule(rule_id, rule)
+        logger.debug(f"List of parsed ApplyRules:\n{self._rules}")
+
+    def add_rule(self, rule_id: int, rule: ApplyRule) -> None:
+        # A common case when a PO subclasses another one with the same URL
+        # pattern. See the ``test_item_return_subclass()`` test case in
+        # ``tests/test_web_poet_rules.py``.
+        matched = self.item_matcher[rule.to_return]
+        pattern_dupes = [
+            pattern
+            for pattern in matched.patterns.values()
+            if pattern == rule.for_patterns
+        ]
+        if pattern_dupes:
+            rules = [
+                r
+                for p in pattern_dupes
+                for r in self.search(for_patterns=p, to_return=rule.to_return)
+            ]
+            warn(
+                f"Similar URL patterns {pattern_dupes} were declared earlier "
+                f"which uses to_return={rule.to_return}. This earlier rule "
+                f"will be used when matching against URLs. Consider updating "
+                f"the priority of these rules: {rules}."
+            )
+
+        if rule.instead_of:
+            self.overrides_matcher[rule.instead_of].add_or_update(
+                rule_id, rule.for_patterns
+            )
+        if rule.to_return:
+            self.item_matcher[rule.to_return].add_or_update(rule_id, rule.for_patterns)
+
+    # TODO: These URL matching functionalities could be moved to web-poet.
+    # ``overrides_for`` in web-poet could be ``str`` or ``_Url`` subclass.
```

Comment:
Thoughts:
- `OverridesRegistryBase` won't be needed if the `overrides_for` functionality is moved to **web-poet**.
- We can immediately retire `OverridesRegistryBase` from **scrapy-poet** if that's the case.

Reply:
I'm considering on moving this functionality right away in **web-poet** so that **scrapy-poet**'s next release already has all of the essential registry functionalities out of the box.

Reply:
Sounds good to me.

Reply:
https://github.com/scrapinghub/web-poet/pull/112